src/securetty.8
.\" Copyright 2009 Petr Lautrbach (plautrba@redhat.com)
.TH securetty 8  2009-10-27 "" "System Administration tools and Daemons"
.SH NAME
securetty \- add a tty to /etc/securetty
.SH SYNOPSIS
.B securetty \fItty\fP
.SH DESCRIPTION
.B securetty
safely adds \fItty\fP to /etc/securetty.
.SH NOTES
.B securetty
is not normally meant to be run by hand; it is invoked by the upstart job that
sets up the system serial console.
.SH "SEE ALSO"
.BR securetty(5)


ChangeLog
commit e2cd6a6b94e6993f4b5ce982c329fecd41efe0fb
Author: Harald Hoyer <harald@redhat.com>
Date:   Tue Jun 16 12:59:27 2009 +0200

    cleanup /var/run/libvirt
    
    https://bugzilla.redhat.com/show_bug.cgi?id=505600
    
    libvirt-0.6.3-8.el5 now includes /var/run/libvirt/network and
    /var/run/libvirt/qemu so the /var cleanup script rc.sysinit needs to be
    extended to accommodate the additional directory depth, otherwise the
    user gets messages on boot about not being able to remove the files because
    they are directories.

 rc.d/rc.sysinit |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

commit 37f574511902944a19a75aa9f30e6263f7702d56
Author: Harald Hoyer <harald@redhat.com>
Date:   Thu May 14 19:57:05 2009 +0200

    version 8.45.28

 ChangeLog        |   24 ++++++++++++++++++++++--
 initscripts.spec |   10 +++++++---
 2 files changed, 29 insertions(+), 5 deletions(-)

commit 960cd8a40e0c73257bd5b302b4718eccc849f004
Author: Harald Hoyer <harald@redhat.com>
Date:   Wed May 6 11:25:39 2009 +0200

    version 8.45.27

 ChangeLog        |   11 +++++++++++
 initscripts.spec |    7 ++++++-
 2 files changed, 17 insertions(+), 1 deletions(-)

commit af6afcb2c78a7f034f2f4dca2310ccf3b575f840
Author: Harald Hoyer <harald@redhat.com>
Date:   Thu May 7 14:12:55 2009 +0200

    fixed "Determine reboot/halt via existing INIT_HALT environment variable. (#475006)"

 rc.d/init.d/halt |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

commit da1f0a03d1a9ae65f36102e71120a3cfe7dd432c
Author: Bill Nottingham <notting@redhat.com>
Date:   Tue Oct 12 01:00:13 2004 +0000

    return of rawdevices stuff

 ChangeLog              |   10 ++++++
 initscripts.spec       |    5 ++-
 rc.d/init.d/rawdevices |   80 ++++++++++++++++++++++++++++++++++++++++++++++++
 sysconfig/rawdevices   |    8 +++++
 4 files changed, 102 insertions(+), 1 deletions(-)

commit 063bde81b8fcd430aa68312c80a514717b6ccd71
Author: Harald Hoyer <harald@redhat.com>
Date:   Tue May 5 11:19:13 2009 +0200

    version 8.45.26

 ChangeLog        |   92 +++++++++++++++++++++++++++++++++++-------------------
 initscripts.spec |    8 +++-
 2 files changed, 66 insertions(+), 34 deletions(-)

commit c03da3323fb88bce8f4978a8a4ca849b1f31298d
Author: Harald Hoyer <harald@redhat.com>
Date:   Mon May 4 16:19:16 2009 +0200

    specifiy mode for O_CREAT

 src/rename_device.c |    3 ++-
 1 files changed, 2 insertions(+), 1 deletions(-)

commit ac34bb846b983eee9f59408ef635366ad618f2e7
Author: Harald Hoyer <harald@redhat.com>
Date:   Mon May 4 16:08:11 2009 +0200

    ignore .git bash files for check target

 Makefile |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

commit eed90b57d56c987e1b5a9135f6a43741c2a596bf
Author: Harald Hoyer <harald@redhat.com>
Date:   Mon May 4 16:01:55 2009 +0200

    changelog and specfile update

 ChangeLog        |  433 +++++++++++++++++++++++++++++++++++++++++++++++++++++-
 initscripts.spec |   35 +++++-
 2 files changed, 459 insertions(+), 9 deletions(-)

commit fb6ab77065b501b2cc5c59be459fa1410b09a833
Author: Harald Hoyer <harald@redhat.com>
Date:   Mon May 4 15:45:54 2009 +0200

    add md activation before lvm/device-mapper in netfs (#480627)
    
    add md activation before lvm/device-mapper in netfs, as is currently done in
    rc.sysinit

 rc.d/init.d/netfs |    3 +++
 1 files changed, 3 insertions(+), 0 deletions(-)

commit 0a207e16a8e8660fdd009352c022e3fb81764b35
Author: Harald Hoyer <harald@redhat.com>
Date:   Mon May 4 15:23:17 2009 +0200

    fixed git commands

 Makefile |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

commit 4442f115105e0b49ce8fa5f245826eb97951b1f3
Author: Harald Hoyer <harald@redhat.com>
Date:   Mon May 4 15:17:06 2009 +0200

    fixed SUBCHANNEL grep for s390

 sysconfig/network-scripts/network-functions |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

commit ee709ad88ac3351ebb833138566599568ca55e8d
Author: Harald Hoyer <harald@redhat.com>
Date:   Mon May 4 15:02:34 2009 +0200

    Move the bridge member handling later in the ifdown sequence (#463325, Sean E. Millichamp <sean@bruenor.org>)

 sysconfig/network-scripts/ifdown-eth |   22 +++++++++++-----------
 1 files changed, 11 insertions(+), 11 deletions(-)

commit 9707e562c1c1e46d55a320aeefff0c64c667f1b1
Author: Bill Nottingham <notting@redhat.com>
Date:   Thu Apr 9 10:29:27 2009 -0400

    Allow changing of VLAN type even if the module is already loaded. (#495053, <pietro@bertera.it>)

 sysconfig/network-scripts/ifup |   12 ++++++------
 1 files changed, 6 insertions(+), 6 deletions(-)

commit d2a76a3ab0931fae034b3992a5068450971b7c22
Author: Bill Nottingham <notting@redhat.com>
Date:   Fri Jul 20 20:33:14 2007 +0000

    add gfs2 to rc.sysinit mount (#494963)

 rc.d/rc.sysinit |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

commit 9e0bfd1047847acc726c63a4026e8710e85f7c71
Author: Bill Nottingham <notting@redhat.com>
Date:   Mon Apr 16 23:47:27 2007 +0000

    support ETHTOOL_OPTS on bridge devices (#493255, <bbaetz@acm.org>)

 sysconfig/network-scripts/ifup-eth |    3 +++
 1 files changed, 3 insertions(+), 0 deletions(-)

commit efd768e3941b54c8e3332dfd96fefed53a4be20e
Author: Bill Nottingham <notting@redhat.com>
Date:   Tue Mar 17 10:16:30 2009 -0400

    Add vlan support for s390 HSI interfaces. (#490584)

 sysconfig/network-scripts/ifup |    3 ++-
 1 files changed, 2 insertions(+), 1 deletions(-)

commit 781b334da9721e37fa75a1d3ea415774867d32e2
Author: Bill Nottingham <notting@redhat.com>
Date:   Fri Feb 6 15:37:46 2009 -0500

    Don't re-init an existing device, it causes errors. (#484411, <jpayne@redhat.com>)

 src/ccw_init |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

commit 4ebc49e3c1720fe8a17286a4fbb2d9ac71b94bb6
Author: Bill Nottingham <notting@redhat.com>
Date:   Tue Mar 17 12:49:42 2009 -0400

    Remove arp_ip_target on ifdown for bonding devices. (#483711)

 sysconfig/network-scripts/ifdown-eth |    4 +++-

src/securetty.c
/*
 * Copyright (c) 2008 Red Hat, Inc. All rights reserved.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA
 *
 */

#include <errno.h>
#include <fcntl.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <syslog.h>
#include <unistd.h>

#include <sys/types.h>
#include <sys/stat.h>

void alarm_handler(int num) {
        return;
}

int open_and_lock_securetty() {
        int fd;
        struct flock lock;
        struct sigaction act, oldact;
        
        lock.l_type = F_WRLCK;
        lock.l_whence = SEEK_SET;
        lock.l_start = 0;
        lock.l_len = 0;
      
        fd = open("/etc/securetty", O_RDWR);
        if (fd == -1) {
                syslog(LOG_ERR, "Couldn't open /etc/securetty: %s",strerror(errno));
                return -1;
        }
        act.sa_handler = alarm_handler;
        act.sa_flags = 0;
        sigaction(SIGALRM, &act, &oldact);
        alarm(2);
        while (fcntl(fd, F_SETLKW, &lock) == -1) {
                if (errno == EINTR) {
                        syslog(LOG_ERR, "Couldn't lock /etc/securetty: Timeout exceeded");
                } else {
                        syslog(LOG_ERR, "Couldn't lock /etc/securetty: %s",strerror(errno));
                }
                return -1;
        }
        alarm(0);
        sigaction(SIGALRM, &oldact, NULL);
        return fd;
}

int rewrite_securetty(char *terminal) {
        int fd;
        char *buf, *pos;
        struct stat sbuf;
        
        fd = open_and_lock_securetty();
        if (fd == -1)
                return 1;
        if (fstat(fd, &sbuf) == -1) {
                close(fd);
                syslog(LOG_ERR, "Couldn't stat /etc/securetty: %s",strerror(errno));
                return 1;
        }
        buf = malloc(sbuf.st_size + 1);
        if (read(fd, buf, sbuf.st_size) != sbuf.st_size) {
                close(fd);
                syslog(LOG_ERR, "Couldn't read /etc/securetty: %s",strerror(errno));
                return 1;
        }
        if (!strncmp(buf,terminal,strlen(terminal)) && buf[strlen(terminal)] == '\n')
                goto out_ok;
        if ((pos = strstr(buf, terminal))) {
                if (pos[strlen(terminal)] == '\n' && *(pos-1) == '\n')
                        goto out_ok;
        }
        if (lseek(fd, 0, SEEK_END) == -1) {
                close(fd);
                syslog(LOG_ERR, "Couldn't seek to end of /etc/securetty: %s",strerror(errno));
                return 1;
        }
        write(fd, terminal, strlen(terminal));
        write(fd, "\n", 1);
out_ok:
        close(fd);
        return 0;
}

int main(int argc, char **argv) {
        if (argc < 2 ) {
                fprintf(stderr, "Usage: securetty <device>\n");
                exit(1);
        }
        openlog("securetty", LOG_CONS, LOG_DAEMON);
        return rewrite_securetty(argv[1]);
}
